src/platform/ESP32/DiagnosticDataProviderImpl.cpp

```diff
@@ -287,6 +287,61 @@ void DiagnosticDataProviderImpl::ReleaseNetworkInterfaces(NetworkInterface * net
     }
 }
 
+CHIP_ERROR DiagnosticDataProviderImpl::GetThreadMetrics(ThreadMetrics ** threadMetricsOut)
+{
+#ifdef CONFIG_FREERTOS_USE_TRACE_FACILITY
+    TaskStatus_t * taskStatusArray = nullptr;
+    ThreadMetrics * head           = nullptr;
+    uint32_t arraySize             = uxTaskGetNumberOfTasks();
+
+    taskStatusArray = (TaskStatus_t *) chip::Platform::MemoryAlloc(arraySize * sizeof(TaskStatus_t));
+    if (taskStatusArray == nullptr)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    uint32_t dummyRunTimeCounter;
+    arraySize = uxTaskGetSystemState(taskStatusArray, arraySize, &dummyRunTimeCounter);
+
+    for (uint32_t i = 0; i < arraySize; i++)
+    {
+        ThreadMetrics * thread = (ThreadMetrics *) chip::Platform::MemoryAlloc(sizeof(ThreadMetrics));
+        if (thread == nullptr)
+        {
+            return CHIP_ERROR_NO_MEMORY;
```

Comment:
If memory allocation for a ThreadMetrics object fails, the function returns early without cleaning up already allocated ThreadMetrics nodes. This causes a memory leak.

Reply:
Also leaks the taskStatusArray.